sql_grader/models.py

```diff
@@ -0,0 +1,222 @@
+    verify_query = None
+    answer_result = None
+    is_ordered = True
+
+    # user-scope
+    submission_query = None
+    submission_result = None
+
+    def __init__(
+            self,
+            database,
+            answer_query,
+            verify_query=None,
+            is_ordered=True,
+    ):
+        self.database = database
+        self.is_ordered = is_ordered
+        self.answer_query = answer_query
+        self.verify_query = verify_query
+        self.answer_result, _ = SqlProblem.run_query(
+            database,
+            answer_query,
+            verify_query,
+        )
+
+    def attempt(self, query):
+        self.submission_query = query
+        submission_result, error = SqlProblem.run_query(
+            self.database,
+            query,
+            self.verify_query,
+        )
+        self.submission_result = submission_result
+        # print('submission_result', submission_result)
+        comparison = SqlProblem.compare_rows(
+            self.answer_result,
+            self.submission_result,
+            is_ordered=self.is_ordered,
+        )
+        # print('comparison', comparison)
+        return comparison, error
+
+    @classmethod
+    def create_database(cls, path_sql):
+        """
+        Create a new in-memory database, initialized via SQL script
+
+        This only needs run once, at startup.
+        """
+        with open(path_sql, 'r') as sql:
+            query = sql.read()
+        connection = cls.create_database_from_sql(query)
+        return connection
+
+    @classmethod
+    def create_database_from_sql(cls, query):
+        connection = sqlite3.connect(':memory:', check_same_thread=False)
+        with connection:
+            connection.executescript(query)
+        return connection
+
+    @staticmethod
+    def clone_database(source):
+        """
+        Copy the contents of a source database into a new in-memory database
+
+        This should be run for each request.
+        """
+        destination = sqlite3.connect(':memory:', check_same_thread=False)
+        query = ''.join(line for line in source.iterdump())
+        destination.executescript(query)
+        return destination
+
+    @classmethod
+    def run_query(cls, source, query, verify_query=None):
+        """
+        Execute the provided SQL query against a copy of the database
+        """
+        def run(database, query):
+            result = []
+            message = None
+            with database as connection:
+                try:
+                    for row in connection.execute(query):
+                        # print('row', row)
+                        result.append(row)
+                except Exception as error:
+                    print('ERROR', error)
+                    result = None
+                    message = unicode(error)
+                    # TODO: more robust error handling
+            return result, message
+        database = cls.clone_database(source)
```

Comment:
Note: We always run the user query against a new copy of the in-memory DB.

Reply:
So it's great that every student gets their own in-memory database, and we don't have to worry about SQL injection. But do we have to worry about any kind of escaping out of sqlite3? Does sqlite3 allow you to issue non-SQL commands that say, read something from disk and echo it somewhere?

Reply:
# TL;DR: I haven't been able to find any such explicit capability.

The closest thing I can find is `sqlite`'s special `dot-commands` [1], which do allow for some related functionality (change the working directory, at least). But that is handled exclusively by the `sqlite` _executable_ and _not_ by the `sqlite` _library_.
I did confirm that I could _not_ issue dot-commands via the Python implementation of `executescript` (which is the one integration point where we pass user input) [2].

From the `sqlite` docs [8]:

## SQLite Always Validates Its Inputs

>  SQLite should never crash, overflow a buffer, leak memory, or exhibit any other harmful behavior, even when presented with maliciously malformed SQL inputs or database files. SQLite should always detect erroneous inputs and raise an error, not crash or corrupt memory. Any malfunction caused by an SQL input or database file is considered a serious bug and will be promptly addressed when brought to the attention of the SQLite developers. SQLite is extensively fuzz-tested to help ensure that it is resistant to these kinds of errors.

> Nevertheless, bugs happen. If you are writing an application that sends untrusted SQL inputs or database files to SQLite, there are additional steps you can take to help reduce the attack surface and prevent zero-day exploits caused by undetected bugs. 

It appears that these settings may only be available as compile-time options [9], so I'm not sure which values are used in our Python deploy and/or the feasibility of modifying them.

Perhaps just setting some reasonable max length (could be defaulted and possibly overridden in Studio) accomplishes the 80/20 for us?

That said, with full disclosure, there do appear to have been a few `sqlite` CVEs throughout the years related to RCE [3][4][5][6][7], though it's not clear whether or not they would apply in our use cases (or are any "different" than CVEs detected in other libraries we use (Django, oauth, etc) [10][11]). A common attack mode revolves around first inserting "attack code" into a persistent database, which is then later read out by the unsuspecting application. Others involve injection. Those such cases shouldn't impact us, given we're only using write-once, in-memory databases.

It's possible there are relevant attack vectors I'm not seeing, so I'd appreciate a second set of eyes (and thumbs up!) on this, before proceeding ;)

Thinking of contingencies if we're still concerned about sandboxing this, how does the Code Jail model differ and would it be worth considering an implementation of a Custom Python Grader. It'd be a much less elegant solution, but it seems feasible?

Anyway, take a look, think it over, and let me know if you want to discuss any more of this further in depth.

Thanks!

- [1] https://sqlite.org/cli.html#special_commands_to_sqlite3_dot_commands_
- [2] https://docs.python.org/3.5/library/sqlite3.html#sqlite3.Cursor.executescript
- [3] https://nakedsecurity.sophos.com/2018/12/19/sqlite-creator-fires-back-at-tencents-bug-hunters/
- [4] https://research.checkpoint.com/2019/select-code_execution-from-using-sqlite/
- [5] https://i.blackhat.com/USA-19/Thursday/us-19-Qian-Exploring-The-New-World-Remote-Exploitation-Of-SQLite-And-Curl-wp.pdf
- [6] https://www.blackhat.com/docs/us-17/wednesday/us-17-Feng-Many-Birds-One-Stone-Exploiting-A-Single-SQLite-Vulnerability-Across-Multiple-Software.pdf
- [7] https://www.theregister.co.uk/2019/05/10/sqlite_rce_vuln/
- [8] https://www.sqlite.org/security.html
- [9] https://www.sqlite.org/security.html#untrusted_sql_inputs
- [10] https://www.cvedetails.com/product/16355/Sqlite-Sqlite.html?vendor_id=9237
- [11] https://www.cvedetails.com/product/18211/Djangoproject-Django.html?vendor_id=10199